paypal/driver.py

```diff
@@ -314,53 +326,111 @@ def DoCapture(
         }
 
         parameters.update(kwargs)
-        query_string = self.signature + urllib.urlencode(parameters)
-        response = urllib2.urlopen(self.NVP_API_ENDPOINT, query_string).read()
-        response_tokens = {}
-        for token in response.split('&'):
-            response_tokens[token.split("=")[0]] = token.split("=")[1]
-        for key in response_tokens.keys():
-            response_tokens[key] = urllib.unquote(response_tokens[key])
+        state = self._nvp_request(parameters)
 
-        state = self._get_value_from_qs(response_tokens, "ACK")
-        self.response = response_tokens
-        self.api_response = response_tokens
-
-        if not state in ["Success", "SuccessWithWarning"]:
+        if state not in ["Success", "SuccessWithWarning"]:
             self.doexpresscheckoutpaymenterror = GENERIC_PAYMENT_ERROR
             self.apierror = self._get_value_from_qs(
-                response_tokens, "L_LONGMESSAGE0"
+                self.response, "L_LONGMESSAGE0"
             )
             return False
         return True
 
-    def DoVoid(
+    def DoReferenceTransaction(
         self,
-        authorizationid,
+        referenceid,
+        paymentaction,
+        amount,
+        currency,
         **kwargs
     ):
+        """
+        If you have a Billing Agreement you can do a reference transaction by
+        passing in the agreement as a reference id. You're strongly recommended
+        to supply the IPADDRESS (as a kwarg) to help detect fraud.
+
+        :param referenceid: The Billing Agreement secret, also known as the
+                            BILLINGAGREEMENTID
+        :type referenceid: string
+
+        :param paymentaction: Authorization/Sale/Order
+        :type paymentaction: string
+
+        :param amount: Amount to charge
+        :type amount: string
+
+        :param currency: 3 letter currency code
+        :type currency: string
+
+        :param kwargs: Additional arguments to be sent with this request
+        :type kwards: dict
+
+        :return: False if the transaction is not marked Success or
+                 SuccessWithWarning, True otherwise.
+        :rtype: bool
+        """
+        parameters = {
+            'METHOD': 'DoReferenceTransaction',
+            'REFERENCEID': referenceid,
+            'PAYMENTACTION': paymentaction,
+            'AMT': amount,
+            'CURRENCYCODE': currency,
+        }
+
+        parameters.update(kwargs)
+        state = self._nvp_request(parameters)
+
+        if state not in ["Success", "SuccessWithWarning"]:
```

Comment:
This list is repeated a lot. It should be a tuple and it should be a constant somewhere in the module to keep it DRY.